clouddrift/binning.py

```diff
@@ -1,19 +1,315 @@
+    return output
+
+
+def _binned_max(
+    flat_idx: np.ndarray, n_bins: int, values: np.ndarray, fill_value: float = np.nan
+) -> np.ndarray:
+    """
+    Compute the maximum of values per bin.
+
+    Parameters
+    ----------
+    flat_idx : array-like
+        1D array of bin indices, same shape as values.
+    n_bins: int
+        number of bins
+    values : array-like
+        1D array of data values
+    fill_value : float, optional
+        Value to fill in bins with no data. Default is np.nan.
+
+    Returns
+    -------
+    result : array-like
+        1D array of length n_bins with the maximum per bin
+    """
+    output = np.full(n_bins, -np.inf)
+    np.maximum.at(output, flat_idx, values)
+    output[output == np.inf] = fill_value
+    return output
+
+
+def _binned_apply_func(
+    flat_idx: np.ndarray,
+    n_bins: int,
+    values: np.ndarray | list[np.ndarray],
+    func: Callable[[list[np.ndarray]], float] = np.mean,
+    fill_value: float = np.nan,
+) -> np.ndarray:
+    """
+    Generic wrapper to apply any functions (e.g., percentile) to binned data.
+
+    Parameters
+    ----------
+    flat_idx : array-like
+        1D array of bin indices.
+    n_bins : int
+        Number of bins.
+    values : array-like or list of array-like
+        1D array (univariate) or list of 1D arrays (multivariate) of data values.
+    func : Callable[[list[np.ndarray]], float]
+        Function to apply to each bin. If multivariate, will receive a list of arrays.
+    fill_value : float, optional
+        Value to fill in bins with no data. Default is np.nan.
+
+    Returns
+    -------
+    result : np.ndarray
+        1D array of length n_bins with results from func per bin.
+    """
+    sort_indices = np.argsort(flat_idx)
+    sorted_flat_idx = flat_idx[sort_indices]
+
+    # single or all variables can be passed as input values
+    if is_multivariate := isinstance(values, list):
+        sorted_values = [v[sort_indices] for v in values]
+    else:
+        sorted_values = values[sort_indices]
+
+    unique_bins, bin_starts = np.unique(sorted_flat_idx, return_index=True)
+    bin_ends = np.append(bin_starts[1:], len(sorted_flat_idx))
+
+    result = np.full(n_bins, fill_value)
+    for i, bin_idx in enumerate(unique_bins):
+        if is_multivariate:
+            bin_values = [v[bin_starts[i] : bin_ends[i]] for v in sorted_values]
+        else:
+            bin_values = sorted_values[bin_starts[i] : bin_ends[i]]
+        result[bin_idx] = func(bin_values)
+
+    return result
+
+
 def binned_statistics(
     coords: np.ndarray | list[np.ndarray],
     data: np.ndarray | list[np.ndarray] | None = None,
     bins: int | list = DEFAULT_BINS_NUMBER,
     bins_range: list | None = None,
     dim_names: list[str] | None = None,
     output_names: list[str] | None = None,
+    statistics: str | list | Callable[[np.ndarray], float] = "count",
     zeros_to_nan: bool = False,
+    fill_value: float = np.nan,
```

Comment:
What is this for? Not described in docstrings.

Reply:
removed